src/sentry/static/sentry/app/views/settings/project/filtersAndSampling/rules/index.tsx

```diff
@@ -3,36 +3,118 @@ import styled from '@emotion/styled';
   onEditRule: (rule: DynamicSamplingRule) => () => void;
   onDeleteRule: (rule: DynamicSamplingRule) => () => void;
-  disabled: boolean;
+  onUpdateRules: (rules: Array<DynamicSamplingRule>) => void;
+};
+
+type RulesWithId = Array<DynamicSamplingRule & {id: string}>;
+
+type State = {
+  rules: RulesWithId;
 };
 
-function Rules({rules, onEditRule, onDeleteRule, disabled}: Props) {
-  return (
-    <StyledPanelTable
-      headers={['', t('Event Type'), t('Category'), t('Sampling Rate'), '']}
-      isEmpty={!rules.length}
-      emptyMessage={t('There are no rules to display')}
-    >
-      {rules.map((rule, index) => (
-        <Rule
-          key={index}
-          rule={rule}
-          onEditRule={onEditRule(rule)}
-          onDeleteRule={onDeleteRule(rule)}
-          disabled={disabled}
+class Rules extends React.PureComponent<Props, State> {
+  state: State = {
+    rules: [],
+  };
+
+  componentDidMount() {
+    this.getRules();
+  }
+
+  componentDidUpdate(prevProps: Props) {
+    if (prevProps.rules !== this.props.rules) {
+      this.getRules();
+    }
+  }
+
+  handleUpdateRulesParent() {
+    const {onUpdateRules} = this.props;
+    const {rules} = this.state;
+
+    const reordered = rules.map(({id: _id, ...rule}) => rule);
+    onUpdateRules(reordered);
+  }
+
+  handleUpdateRules = (ruleIds: Array<string>) => {
+    const {rules} = this.state;
+    const reorderedRules = ruleIds
+      .map(ruleId => rules.find(rule => rule.id === ruleId))
+      .filter(rule => !!rule) as RulesWithId;
+
+    this.setState({rules: reorderedRules}, this.handleUpdateRulesParent);
+  };
+
+  getRules() {
+    const {rules} = this.props;
+    const rulesWithId = rules.map((rule, index) => ({...rule, id: String(index)}));
+    this.setState({rules: rulesWithId});
+  }
+
+  render() {
+    const {onEditRule, disabled, onDeleteRule} = this.props;
+    const {rules} = this.state;
+
+    return (
+      <StyledPanelTable
+        headers={['', t('Event Type'), t('Category'), t('Sampling Rate'), '']}
+        isEmpty={!rules.length}
+        emptyMessage={t('There are no rules to display')}
+      >
+        <DragHandle
+          items={rules.map(rule => rule.id)}
+          onUpdateItems={this.handleUpdateRules}
+          renderItem={({
+            value,
+            listeners,
+            forwardRef,
+            transform,
+            transition,
+            style: grabStyle,
+          }) => {
+            const currentRule = rules.find(rule => rule.id === value);
+
+            if (!currentRule) {
+              return null;
+            }
+
+            const style = {
+              transition,
+              '--translate-x': transform ? `${Math.round(transform.x)}px` : undefined,
+              '--translate-y': transform ? `${Math.round(transform.y)}px` : undefined,
+              '--scale-x': transform?.scaleX ? `${transform.scaleX}` : undefined,
+              '--scale-y': transform?.scaleY ? `${transform.scaleY}` : undefined,
```

Comment:
Is this to bypass emotion recalculating CSS rules on each dragmove?

Reply:
yes

Reply:
I like this a lot. Nice work @priscilawebdev